1.48x Faster Partial Caching - Alternative to #36310

schneems opened the pull request:
This is a "safer" but slower alternative to #36310.

```
Then: 5.97/3.29 => 1.81x faster
Now: 5.97/4.02 => 1.48x faster
```

So the change in performance from #36310 to this PR is  about

```
1 - 3.29/4.02 => 18% # slower than #36310 but still 1.48x faster than master.
```

In #36310 performance was the main goal. In this PR the primary goal was to change as little as possible, performance was the secondary goal. The main difference between the two PRs comes from how `cache_version` and `cache_key` are computed. In #36310 that logic was ported over to the `Cache::Key` class where it could be optimized, in this PR the logic was not moved and instead the `Cache::Key` class leverages the `expanded_version` method in their original location.

How does this PR improve performance for rendering a partial and passing in a collection? 

- On master when a key is used in the cache it has to be generated 3 different times. By using the `Cache::Key` class, it is only generated once and then re-used
- On master, the key that represents the view partial that is being rendered is calculated for every object being pulled from cache inside of the loop, in this PR that view key is generated once outside the loop and re-used.

For safety: In addition to not changing the location of the `expanded_version` and `expanded_key` logic, this class will respect any modified logic of the `Rails.cache` store. So if a cache store customizes their version of `expanded_version` then that version will be used when generating the cache.

Here's some raw numbers from the benchmark provided in #36310 when applied to this PR.

```
$ bundle exec ruby perf_scripts/partial_cache.rb
-- create_table(:customers, {:force=>true})
   -> 0.0115s
/Users/rschneeman/Documents/projects/codetriage/perf_scripts/partial_cache
<#ActiveSupport::Cache::MemoryStore entries=1000, size=352893, options={}>
Warming up --------------------------------------
collection render: no cache
                        20.000  i/100ms
collection render: with cache
                         4.000  i/100ms
Calculating -------------------------------------
collection render: no cache
                        199.729  (± 9.0%) i/s -      1.000k in   5.051812s
collection render: with cache
                         49.642  (± 4.0%) i/s -    248.000  in   5.008595s

Comparison:
collection render: no cache:      199.7 i/s
collection render: with cache:       49.6 i/s - 4.02x  slower
```

> Note that while it's still slower, the prior version was 5.97x slower.

schneems committed:
1.8x Faster Partial Caching - Faster Cache Keys

In rails/rails#35257 Aaron pointed out that partial caching is very slow when the partial is very
 small. When I ran the same benchmark locally (without this patch) I get around 5.97x slower than
no caching. With this PR I get 3.29x slower. This represents a 1.8x improvement in speed.

With this PR i'm seeing a 1.09x faster performance improvement in CodeTriage response time
via derailed_benchmarks.

Currently, cache keys are generated via method calls. When a single object is retrieved the raw object that the cache key is built from must be iterated over twice. Once to build the cache key and another for the cache version. In addition to generating the cache key for interacting with the cache, it is also re-generated for logging inside of the `instrument` call. These two things make generating cache keys slow.

To make this faster I moved the key generation to an object. As elements are added to the key object a cache key and cache version are built incrementally. Since we can build both at the same time we can optimize this execution. In this model we only generate cache key once, retrieving the cache key multiple times on this object is now fast.

One other cause of slowness is that when the partial is being rendered the view generates and prepends the "combined_fragment_cache_key" to each of the cache keys. Because our cache objects are now composable we can instead generate the "combined_fragment_cache_key" once and re-use the same object for generating N cache keys.

This is building off the work of rails/rails#33944. The main difference here is that this PR is backwards compatible.

You can run this benchmark by following these instructions:

```
$ git clone https://github.com/codetriage/codetriage
$ cd codetriage
$ git checkout schneems/keymaster-reproduction
$ bundle install
$ bundle exec ruby perf_scripts/partial_cache.rb
-- create_table(:customers, {:force=>true})
   -> 0.0092s
/Users/rschneeman/Documents/projects/codetriage/perf_scripts/partial_cache
<#ActiveSupport::Cache::MemoryStore entries=1000, size=352893, options={}>
Warming up --------------------------------------
collection render: no cache
                        27.000  i/100ms
collection render: with cache
                         7.000  i/100ms
Calculating -------------------------------------
collection render: no cache
                        272.792  (± 2.9%) i/s -      1.377k in   5.052080s
collection render: with cache
                         82.970  (± 2.4%) i/s -    420.000  in   5.066222s

Comparison:
collection render: no cache:      272.8 i/s
collection render: with cache:       83.0 i/s - 3.29x  slower
```

schneems committed:
Do not move location of key generation logic

In the prior commit the location of the key and version generation was moved to a custom object. While faster, it was more confusing to understand if the logic was the same or different.

This PR works by treating the Cache::Key as a value store that cache's the value of cache_key and cache_version so that it only needs to be generated once instead of multiple times.

schneems committed:
Use App's cache store if defined

schneems committed:
Preserve existing interfaces for smaller diff

schneems committed:
Improve readability of logic

schneems committed:
Make code climate happy

schneems committed:
Correct for subtle difference of cache with array

schneems committed:
Faster Rails.cache detection

kaspth commented on activesupport/lib/active_support/cache/key.rb:
```diff
@@ -0,0 +1,153 @@
+      end
+
+      def cache_key_with_version
+        @cache_method_obj.send(:retrieve_cache_key, @key_parts)
```
Hm, the `send`s in here bother me, but I think we can live with them if the future design idea is to flip the relationship such that a cache store really consults the key for `retrieve_cache_key` etc.

schneems replied:
> if the future design idea is to flip the relationship such that a cache store really consults the key for retrieve_cache_key etc.

Inverting that relationship is the core idea with #36310 where the key logic would live in the object rather than in external recursive methods (i.e. `normalize_key` and `normalize_version`). It might be worth re-visiting it now if the eventual goal would be to get there. I could make it less invasive and restrict the key generation to only view collection caching. 

> Hm, the sends in here bother me

I agree that the sends here are unnerving. The interface provided by Cache::Store is a bit odd since most people who implement a cache store inherit from this store meaning they can freely use these methods even though they're written as being private. I could make them public (because in a very real way they're already public today) however that wouldn't fix any issues with other cache stores that have implemented their own methods and decided to keep them private.

This also does highlight probably the most "unsafe" thing about this PR versus #36310, it assumes any cache store provided has these 3 methods on them. To my knowledge, the vast majority of people use the Rails provided cache stores and some use `:dalli`. I'm not sure if there are any other cache stores to consider.

kaspth reviewed (commented):
Hey Richard, really cool that you went the extra length of restructuring to slim your PR, especially when we we're negligent in reviewing. I'm sorry it took so long and thank you for your patience ❤️🙏

Though now I'm throwing a lot at you, so if you'd rather that I address my concerns in commits after this PR, then I'll happily merge now and look into them. Or we can do a bit of both, or just the usual, your pick 🤠

schneems committed:
Reduce docs

schneems committed:
Don't require ActiveSupport::Cache::Store at boot

schneems commented on activesupport/lib/active_support/cache/key.rb:
```diff
@@ -0,0 +1,153 @@
+      end
+      alias :<< :update
+
+      def length
```
I needed this in #36310 due to some tests that were asserting based on key objects since in that PR the Cache::Key objects ended up being passed around instead of only in this one collection cache context. 

I think we don't need it in this PR, I'll remove it and watch CI.

schneems replied:
CI works with this code removed, going to resolve this comment.

schneems committed:
Remove un-needed Cache::Key#length method

schneems commented on activesupport/lib/active_support/cache/key.rb:
```diff
@@ -0,0 +1,153 @@
+        @cache_version ||= @cache_method_obj.send(:expanded_version, @key_parts)
+
+        if @cache_version == "/"
+          nil
```
Without this code we get three failures in the tests:

```
CachedCollectionViewRenderTest#test_collection_caching_with_partial_that_doesn't_use_fragment_caching [/Users/rschneeman/Documents/projects/rails/actionview/test/template/render_test.rb:797]:
Expected: "Cached"
  Actual: "Hello: david"


rails test Users/rschneeman/Documents/projects/rails/actionview/test/template/render_test.rb:791

.................................F

Failure:
CachedCollectionViewRenderTest#test_collection_caching_with_cached_true [/Users/rschneeman/Documents/projects/rails/actionview/test/template/render_test.rb:807]:
--- expected
+++ actual
@@ -1 +1,2 @@
-"Cached"
+"  Hello: david
+"



rails test Users/rschneeman/Documents/projects/rails/actionview/test/template/render_test.rb:801

..F

Failure:
CachedCollectionViewRenderTest#test_template_body_written_to_cache [/Users/rschneeman/Documents/projects/rails/actionview/test/template/render_test.rb:778]:
Expected: "Hello: david"
  Actual: nil


rails test Users/rschneeman/Documents/projects/rails/actionview/test/template/render_test.rb:773
```

Here's an example of one of the tests:

```ruby
  test "collection caching with cached true" do
    customer = CachedCustomer.new("david", 1)
    key = cache_key(customer, "test/_cached_customer")

    ActionView::PartialRenderer.collection_cache.write(key, "Cached")

    assert_equal "Cached",
      @view.render(partial: "test/cached_customer", collection: [customer], cached: true)
  end
```

The issue at heart is that `nil` and empty values weren't really taken into account when the current `noramlize_version` was written so the behavior is somewhat undefined (or rather it's unclear what behavior we _want_).

For example if you try to generate a cache version based off of a symbol you'll get a nil, but if you try to generate it off of a symbol in an array you'll get an empty string:

```ruby
ActiveSupport::Cache::MemoryStore.new.send(:normalize_version, :views)
=> nil
ActiveSupport::Cache::MemoryStore.new.send(:normalize_version, [:views])
=> ""
```

These tests are failing because of the base assumption that normalize version will treat `[key]` the same as `key`. This is mostly true, but not in the case of empty values. In these failures we're generating a cache_version when we wouldn't have originally

Basically the difference between:

```
ActiveSupport::Cache::MemoryStore.new.send(:normalize_version, [[:views], "value", []])
# => "/"
```

and 

```
ActiveSupport::Cache::MemoryStore.new.send(:normalize_version, [:views, "value", []])
# => ""
```

My initial solution was to consider "/" the same as an empty string. I'm looking into seeing if there's another way to generate or insert keys that doesn't require this explicit knowledge.

schneems replied:
I've got a solution here https://github.com/rails/rails/pull/37301/commits/a01a94568cd38ceb40bd27d5ef8771565607b193

Essentially we can avoid this issue by making sure that items are directly added to the `Cache::Key`, so rather than adding a whole array to the key, we can add it's constituent parts. This method has the benefit of removing an array allocation from the internal loop.

schneems committed:
Do not re-nest fragment cache key

Based on this comment https://github.com/rails/rails/pull/37301#discussion_r329372520.

Currently arrays are treated like cache key containers. For the purposes of generating keys `[:foo]` will generate the same key as `:foo` without the array. Unfortunately there's an inconsistency with the way that cache versions are generated so that arrays are not perfect cache containers:

```ruby
ActiveSupport::Cache::MemoryStore.new.send(:normalize_version, :views)
=> nil
ActiveSupport::Cache::MemoryStore.new.send(:normalize_version, [:views])
=> ""
```

```
ActiveSupport::Cache::MemoryStore.new.send(:normalize_version, [[:views], "value", []])
# => "/"
ActiveSupport::Cache::MemoryStore.new.send(:normalize_version, [:views, "value", []])
# => ""
```

My initial solution was to consider "/" the same as an empty string.

This PR instead works by inserting the individual key parts of the `combined_fragment_cache_key` into our generated Cache::Key object which means that our internal array `@key_parts` in the Cache::Key object will contain the same nesting structure originally used.

In addition I've added an interface in the `cache_helper.rb` file to allow me to pass in a container object such as an array or a Cache::Key object. This reduces the number of allocations needed and prevents un-needed nesting of arrays which might create the cache version problem.

schneems commented on activesupport/lib/active_support/cache.rb:
```diff
@@ -851,3 +851,4 @@ def uncompress(value)
     end
   end
 end
+require "active_support/cache/key"
```
Good point. I fixed this so the key is lazily generated and memoized to the class in a way that doesn't require this circular dependency.

schneems commented on activesupport/lib/active_support/cache/key.rb:
```diff
@@ -0,0 +1,153 @@
+    #
+    #   klass = Class.new { def cache_version; "foo"; end }
+    #   Key.new({foo: klass.new}).cache_version
+    #   # => "foo"
```
I reduced the docs to one example. People can view the tests for more if the want/need. These docs are less needed as this class is much smaller than #36310, it's easier to grok the purpose of this class at a glance versus that PR.

schneems commented on activesupport/lib/active_support/cache/key.rb:
```diff
@@ -0,0 +1,153 @@
+        @cache_method_obj = cache_method_obj
+
+        self << key unless key == DEFAULT_KEY
+      end
```
I realized that we're always initializing with a "base" key so I changed it to a required argument and avoided calling the `<<` method altogether. As the class is internal only for now (:nodoc:) I think it's okay for the object to be more opinionated about how it's used. for now. Let me know if you disagree or you're worried. I could add a builder method. 

I do like the way that `Key.chain(thing)` reads in code.

schneems committed:
Fix test to return same type as original interface

schneems commented on activesupport/lib/active_support/cache/key.rb:
```diff
@@ -0,0 +1,153 @@
+          else
+            CACHE_METHOD_OBJ
+          end
+        end
```
I like using dependency injection, switched to an optional second argument that defaults to grabbing the information from the class.

I tried to inject it in `collection_caching.rb` directly but when I do I get several failures since Rails isn't guaranteed to exist and have a `cache` object on it in the actionview library

```
Error:
CachedCollectionViewRenderTest#test_collection_caching_with_partial_that_doesn't_use_fragment_caching:
NoMethodError: undefined method `cache' for Rails:Module
```

kaspth replied:
You shouldn't call `Rails.cache` there, we've already set up a `collection_cache` to use there which fetches what's basically `Rails.cache` via the Railtie: https://github.com/rails/rails/blob/042b6fec877582f13775903f64d2c5e1da485803/actionview/lib/action_view/railtie.rb#L96

We should convert this to `mattr_accessor` ala https://github.com/rails/rails/blob/042b6fec877582f13775903f64d2c5e1da485803/actionview/lib/action_view/renderer/partial_renderer/collection_caching.rb#L10 and remove the default_cache_object method.

schneems committed:
Use dependency injection instead of calling class

schneems committed:
Namespace Cache::Key tests

schneems commented on activesupport/test/cache/key_test.rb:
```diff
@@ -0,0 +1,100 @@
+require "abstract_unit"
+require "active_support/cache"
+
+class KeyTest < ActiveSupport::TestCase
```
Good idea. There's a lot of other tests inside of the `activesupport/cache` folder that could benefit from this pattern. I updated my PR to implement your suggestion. Resolving this comment.

kaspth commented on activesupport/lib/active_support/cache/key.rb:
```diff
@@ -0,0 +1,60 @@
+      end
+
+      def cache_key_with_version
+        @cache_object.send(:retrieve_cache_key, @key_parts)
```
Not too keen on the `_object` variable suffix since everything is an object in Ruby, so it doesn't signify much 😄 How about `@cache_store`?

kaspth commented on actionview/lib/action_view/renderer/partial_renderer/collection_caching.rb:
```diff
@@ -59,17 +59,23 @@ def collection_by_cache_keys(view, template)
+          key = ActiveSupport::Cache::Key.new(base)
+          view_parts.each do |part|
+            key << part
+          end
```
Could add a `concat` method to `Key` that delegates to `concat` on `@key_parts` and do `key.concat view_parts` should be clearer and quicker still.

kaspth commented on actionview/lib/action_view/helpers/cache_helper.rb:
```diff
@@ -227,16 +228,22 @@ def digest_path_from_template(template) # :nodoc:
         else
+          container << name if container
           name
         end
```
Don't think this is worth saving the extra array allocation. Curious if we could replace all this with something like:

```ruby
view.cache_fragment_name(item, virtual_path: template.virtual_path, digest_path: digest_path).tap do |key_parts|
  key.concat Array(key_parts)
end
```

in the collection caching stuff. Fine if you want to replace `Array()` with more granular handling.

kaspth reviewed (commented):
Getting closer!

schneems commented:
Thanks for the comments I've not been able to re-visit this PR.

As a data point, I did some benchmarking with Derailed against CodeTriage and I was able to make the homepage 1.5x faster by disabling collection cache. I don't think even my patches would have been enough to break even since the partials were so small https://github.com/codetriage/codetriage/pull/1226.

While I do think we can work to make the collection cache code faster, i'm also thinking that the bigger problem people are likely having is that they're using collection caching for tiny partials when they shouldn't.
